### PR TITLE
Mean vertex calibration

### PR DESCRIPTION
**Pull request by chiarazampolli**
This is the calibration for the mean vertex.
For now, it was tested running on simulated reconstructed primary vertices, so not yet in the form of "TF-based" processing, but reading from a file.

To test, there are 2 ways:

o2-primary-vertex-reader-workflow -b | o2-calibration-mean-vertex-calibration-workflow -b --> it does the fit only on the merged slots (merged to do the SMA)

o2-primary-vertex-reader-workflow -b | o2-calibration-mean-vertex-calibration-workflow --useFit -b --> it does the fit at every slot, and then does the SMA on the fits

### CODE REVIEW

File: DataFormats/Calibration/src/MeanVertexObject.cxx
Comment (shahor02) [inline DataFormats/Calibration/src/MeanVertexObject.cxx]: actually, the default copy constructor would do the same.

File: DataFormats/Calibration/src/MeanVertexObject.cxx
Comment (shahor02) [inline DataFormats/Calibration/src/MeanVertexObject.cxx]: idem

File: DataFormats/Calibration/src/MeanVertexObject.cxx
Comment (shahor02) [inline DataFormats/Calibration/src/MeanVertexObject.cxx]: idem

File: DataFormats/Calibration/src/MeanVertexObject.cxx
Comment (shahor02) [inline DataFormats/Calibration/src/MeanVertexObject.cxx]: idem

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (shahor02) [inline Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: +- 1 cm in XY should be enough (and current 200 micron resolution is too crude). In Z the range should be ~+-20.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (shahor02) [inline Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: I think all these deques (also those below) should be a ``deque<XXX>``, where XXX is a struct derived from MeanVertexData + the TF start/end
like now, it will require 8 memory allocations instead of 1 at deque expanding.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (shahor02) [inline Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Also here: they all should be processed in one go, why not using ``MeanVertexObject``?

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h
Comment (shahor02) [inline Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h]: here also better to use automatically generated move c-tor.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h
Comment (shahor02) [inline Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h]: idem here and below.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h
Comment (shahor02) [inline Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h]: I don't understand, why do you modify ``other``? If to reset it, one should rely on default assignment operator and add separate ``clear`` method.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h
Comment (shahor02) [inline Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h]: idem. Also, for move c-tor there is no point in resetting the original object.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h
Comment (shahor02) [inline Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h]: why deque and not a vector?

File: Detectors/Calibration/src/MeanVertexData.cxx
Comment (shahor02) [inline Detectors/Calibration/src/MeanVertexData.cxx]: it is faster to do e.g.
````
uint32_t binx = (x+rangeX)*v2BinX;
uint32_t biny = (y+rangeY)*v2BinY;
uint32_t binz = (z+rangeZ)*v2BinZ;
if (binx<nbinsX || biny<nbinsY || binz<nbinsZ) { // accounts also for z<-rangeZ
  histoX[binx]++;
  histoY[biny]++;
  histoZ[binz]++;
}
````

File: Detectors/Calibration/src/MeanVertexData.cxx
Comment (shahor02) [inline Detectors/Calibration/src/MeanVertexData.cxx:86]: Why do you impose same nbins for all directions? We may have different nbins for XY and Z, better use proper size() for each of them.

File: Detectors/Calibration/src/MeanVertexData.cxx
Comment (shahor02) [inline Detectors/Calibration/src/MeanVertexData.cxx]: idem

File: Detectors/Calibration/workflow/src/MeanVertexCalibratorSpec.cxx
Comment (shahor02) [inline Detectors/Calibration/workflow/src/MeanVertexCalibratorSpec.cxx]: I think this setting better to pass via ConfigurableObject, which will be eventually interfaced to ccdb.

**Review by shahor02 — COMMENTED** [review]
Hi @chiarazampolli 
Please see some comments below.
Cheers,
 Ruben

File: DataFormats/Calibration/src/MeanVertexObject.cxx
Comment (chiarazampolli) [reply DataFormats/Calibration/src/MeanVertexObject.cxx]: Yes, I implemented all these because I wanted to use std::move. I can change to setting it "= default", if you prefer.

File: DataFormats/Calibration/src/MeanVertexObject.cxx
Comment (chiarazampolli) [reply DataFormats/Calibration/src/MeanVertexObject.cxx]: I did not know I could use the "= default" for this.

File: DataFormats/Calibration/src/MeanVertexObject.cxx
Comment (chiarazampolli) [reply DataFormats/Calibration/src/MeanVertexObject.cxx]: As for the copy ctor.

File: DataFormats/Calibration/src/MeanVertexObject.cxx
Comment (chiarazampolli) [reply DataFormats/Calibration/src/MeanVertexObject.cxx]: As for the move ctor.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: These values are not to be considered, because it is the device that sets them. But I can change it to avoid confusion. 
The units are cm, right? At least from what I saw (I think I did not see in the PrimaryVertex and Vertex class). From the simulated events, 0.1 was enough for X and Y.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h]: It was a vector initially, I changed when I wanted to implement the "subtract" method, but it is not needed. I will put back vector.

File: Detectors/Calibration/src/MeanVertexData.cxx
Comment (chiarazampolli) [reply Detectors/Calibration/src/MeanVertexData.cxx]: Will change it, thanks!

File: Detectors/Calibration/src/MeanVertexData.cxx
Comment (chiarazampolli) [reply Detectors/Calibration/src/MeanVertexData.cxx:86]: This is indeed wrong, inherited from my first work on the code.

File: Detectors/Calibration/src/MeanVertexData.cxx
Comment (chiarazampolli) [reply Detectors/Calibration/src/MeanVertexData.cxx]: yes.

File: Detectors/Calibration/workflow/src/MeanVertexCalibratorSpec.cxx
Comment (chiarazampolli) [reply Detectors/Calibration/workflow/src/MeanVertexCalibratorSpec.cxx]: Ok, I will take a look on how to use it.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Updated (assuming cm), also the binning (as I said, the device is set differently, but better have meaningful ranges and bins). Can you check?

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: I thought about this option, I thought this was was easier for the moving average, I will change it. I will then change the `std::deque<MeanVertexData> mTmpMVdataDq;` to be also a deque of this struct (but I will do this next week).

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Ok, I will change this, and also the `doSimpleMovingAverage` method, of course.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h]: Done

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h]: I thought that this was needed to call std::move correctly. 
I now use the default implementations, so I removed it.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexData.h]: Idem, default used now. If it is ok to rely on that for the std::move, then it is fine with me.

File: DataFormats/Calibration/src/MeanVertexObject.cxx
Comment (shahor02) [reply DataFormats/Calibration/src/MeanVertexObject.cxx]: I would define them ``= default`` only if compiler requires it, otherwise it would be generated automatically.

File: DataFormats/Calibration/src/MeanVertexObject.cxx
Comment (shahor02) [reply DataFormats/Calibration/src/MeanVertexObject.cxx]: I think you can leave it to compiler.

Comment (chiarazampolli) [reply DataFormats/Calibration/src/MeanVertexObject.cxx]: Isn't it that if you don't implement them, then the std::move is actually a copy?

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (shahor02) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: yes, all units are cm. But in the constructor I still see old defaults.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Hi @shahor02 ,
I was thinking about modifying the MeanVertexData and adding there as members the start and end, in general. What do you think? 
And: What do you think about changing the MeanVertexObject from class to struct?
Chiara

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Done. Sorry that I had forgotten.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Actually I have a second thought. I think the time should stay separate, because I use it in both cases:
- when I fit the single slot and then do the average of the results of the fits;
- when I sum up the slots, and then do the fit. 
So I will indeed change that we do not allocate one deque per data member of the MeanVertexObject, but the time (start and end) will stay separate. Let's see how it looks.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Implemented, can you please check?

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Changed, can you check?

File: Detectors/Calibration/workflow/src/MeanVertexCalibratorSpec.cxx
Comment (chiarazampolli) [reply Detectors/Calibration/workflow/src/MeanVertexCalibratorSpec.cxx]: Ok, I gave it a try, if you can check, it is this commit: 5361947b6e4bcfc633631de79a4ea05be43e1811

File: Detectors/Calibration/src/MeanVertexData.cxx
Comment (shahor02) [inline Detectors/Calibration/src/MeanVertexData.cxx]: sorry for bad suggestion, actually, due to the rounding one has to do
````
auto dx = x+rangeX;
uint32_t binx = dx<0 ? 0xffffffff : (x + rangeX) * v2BinX;
...
````
BTW, I am preparing now flat histo class for VDrift calibration, can be eventually used also here.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (shahor02) [inline Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Actually, better use std::deque<o2::math_utils::Bracket<TFType>>, to avoid 2 reallocations.
But I think your 1st idea is better: you can define the time range in the MeanVertexData itself.
I did not understand your arguments why this will not work.

**Review by shahor02 — COMMENTED** [review]
Hi @chiarazampolli 

The ``MeanVertexParam`` is looks fine. See 2 comments below.

Cheers,
 Ruben

File: Detectors/Calibration/src/MeanVertexData.cxx
Comment (chiarazampolli) [reply Detectors/Calibration/src/MeanVertexData.cxx]: Thanks! I had copied and pasted blindly :-)

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: The point is that mTmpMVobjDqTimeStart is used to define the start validity of the CCDB object. If I bound it to the MeanVertexData itself, when I do the SMA of the fits, I will not have this available, since the case "mUseFit" fills a deque of MeanVertexObjects (no data) on which then I will then do the SMA. 
BTW, I think that the way the `startValidity` is defined is wrong. Currently it just does the average of all start times of the various slots, but this will not necessarily be a uint64_t, as required as start validity for a CCDB entry (now we use the TF id, in the future it should become the timestamp).
In addition, when I have enough slots, I would do a SMA also for the start validity. 
The other possibility is to associate as start validity always the start of the first slot used for the average, but this is not what we discussed. 
What do you think?

PS. I removed "mTmpMVobjDqTimeEnd", because this is not needed.

Comment (shahor02) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: OK, I see. For the startValidity, indeed, I did not notice you do this in this way. But in any case, it is uint64_t simply because you convert the mean of the all TimeStarts to that. But I think it is much simpler to simply assign the value to the center of the window, i.e. (mTmpMVobjDqTimeStart.front() + mTmpMVobjDqTimeEnd.back())/2;

I did not understand what do you mean by doing SMA for the startValidity? This is a point in time to which you assign the SMA of vertex, why do you need to apply SMA to it?

Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Hi Ruben, 
That is what I wanted to do, assign in the middle of the window. But I forgot to remove the first element of the deque (for the start time), so I was always taking all start times that had been seen. This is the first problem which I need to fix (and indeed I need the end time too). 
Secondly, since I use uint64_t, dividing by 2 will lead to a sort of rounding effect. If this is not an issue, but a good enough approximation, then fine with me. The other possibility was to do the average of the start times of the slots over which I do the running average - this is what I meant. 
I will implement now as you suggested.
Chiara

Comment (shahor02) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Hi @chiarazampolli 
small rounding is not a problem, the coarsest unit of this uint64_t will be TF (might be eventually an orbit), so the error will be ~5ms, which is irrelevant for the MV variation.

Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Yes, I agree.
Now I fixed the determination of the startValidity. I think nothing else is missing. Right? Aprt for more testing, with more events and TFs.

File: Detectors/Calibration/CMakeLists.txt
Comment (shahor02) [inline Detectors/Calibration/CMakeLists.txt]: can you remove this?

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (shahor02) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: It's ok, but I see you are still using 2 separate dequeus for start and end, leading to 2 reallocations. Why not using a single container with pair of timestamps? The Bracket class I mentioned provides also some comparison operations.

**Comment by chiarazampolli** [comment]
As for the deque with the Bracket. Yes, I can do so. For the end time one does not strictly need a deque, but since I will put it with the start, then it is fine. 
Doing it now.

File: Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h
Comment (chiarazampolli) [reply Detectors/Calibration/include/DetectorsCalibration/MeanVertexCalibrator.h]: Done.

**Comment by chiarazampolli** [comment]
Hello @shahor02 ,
Now I am using Bracket. I don't need (for now) the comparison operators. Let me know if I forgot anything else.
Thanks!
Chiara

**Comment by chiarazampolli** [comment]
Hello @shahor02 ,

Do you want that I change the histograms to use those in https://github.com/AliceO2Group/AliceO2/pull/4693, or can this be merged, and the change to the flat histo I do in another PR (together with the same change for the TOF calibration)?

Chiara

**Comment by shahor02** [comment]
Hi @chiarazampolli,
Let's do this in other PR (perhaps extra methods will need to be added to flat histos).
Could you remove the WIP, I'll merge this once the tests are passed.
Cheers,
 Ruben

**Comment by chiarazampolli** [comment]
Done, thanks!

Chiara

**Review by shahor02 — APPROVED** [review]

**Comment by shahor02** [comment]
@chiarazampolli squashing before merging

**Pull request merged**
